src/System.Runtime.Environment/src/OSName.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Runtime.InteropServices
+{
+    public struct OSName : IEquatable<OSName>
+    {
+        private readonly string _name;
+
+        private const string WindowsName = "WINDOWS";
+        private const string LinuxName = "LINUX";
+        private const string OSXName = "OSX";
+
+        public static readonly OSName Windows = new OSName(WindowsName);
+        public static readonly OSName Linux = new OSName(LinuxName);
+        public static readonly OSName OSX = new OSName(OSXName);
+
+        public OSName(string name)
+        {
+            if (name == null) throw new ArgumentNullException("name");
+            if (name == string.Empty) throw new ArgumentException("name");
```

Comment:
The `ArgumentException` constructor that takes a `string` parameter is actually the `message` parameter, not the `paramName` parameter like with `ArgumentNullException`. Unfortunately, in order to specify the `paramName` with `ArgumentException` it requires specifying a message, which means carrying a new localizable resource for this assembly.

Something like:

```c#
if (name.Length == 0) throw new ArgumentException(SR.Arg_Empty, "name");
```

Where `SR.Arg_Empty` is something like `"Value cannot be empty."`.

Validating empty is such a common operation, I'm going to propose adding a new `ArgumentEmtpyException` along the lines of `ArgumentNullException`, to avoid having to specify a localizable `"Value cannot be empty."` message in each assembly that has to do such empty checks.

Then a localizable resource wouldn't be needed in this assembly for this:

```c#
if (name.Length == 0) throw new ArgumentEmptyException("name");
```